lumigator/python/mzai/backend/backend/services/workflows.py

```diff
@@ -33,60 +31,79 @@ def __init__(
         dataset_service: DatasetService,
+        tracking_client: TrackingClient,
     ):
         self._job_repo = job_repo
         self._job_service = job_service
         self._dataset_service = dataset_service
+        self._tracking_client = tracking_client
+        self.NON_TERMINAL_STATUS = [
+            JobStatus.CREATED.value,
+            JobStatus.PENDING.value,
+            JobStatus.RUNNING.value,
+        ]
 
-    NON_TERMINAL_STATUS = [
-        JobStatus.CREATED.value,
-        JobStatus.PENDING.value,
-        JobStatus.RUNNING.value,
-    ]
-    """list: A list of non-terminal job statuses."""
+        # TODO: rely on https://github.com/ray-project/ray/blob/7c2a200ef84f17418666dad43017a82f782596a3/python/ray/dashboard/modules/job/common.py#L53
+        self.TERMINAL_STATUS = [JobStatus.FAILED.value, JobStatus.SUCCEEDED.value]
 
-    # TODO: rely on https://github.com/ray-project/ray/blob/7c2a200ef84f17418666dad43017a82f782596a3/python/ray/dashboard/modules/job/common.py#L53
-    TERMINAL_STATUS = [JobStatus.FAILED.value, JobStatus.SUCCEEDED.value]
-    """list: A list of terminal job statuses."""
-
-    async def on_job_complete(self, job_id: UUID, task: Callable = None, *args):
-        """Watches a submitted job and, when it terminates successfully, runs a given task.
-
-        Inputs:
-        - job_id: the UUID of the job to watch
-        - task: the function to be called after the job completes successfully
-        - args: the arguments to be passed to the function `task()`
-        """
-        job_status = self._job_service.get_upstream_job_status(job_id)
-
-        loguru.logger.info(f"Watching {job_id}")
-        while job_status not in self.TERMINAL_STATUS and job_status in self.NON_TERMINAL_STATUS:
-            await asyncio.sleep(5)
-            job_status = self._job_service.get_upstream_job_status(job_id)
-
-        match job_status:
-            case JobStatus.FAILED.value:
-                loguru.logger.error(f"Job {job_id} failed: not running task {str(task)}")
-            case JobStatus.SUCCEEDED.value:
-                loguru.logger.info(f"Job {job_id} finished successfully.")
-                if task is not None:
-                    task(*args)
-            case _:
-                # NOTE: Consider raising an exception here as we *really* don't expect
-                # anything other than FAILED or SUCCEEDED.
-                loguru.logger.error(
-                    f"Job {job_id} has an unexpected status ({job_status}) that is not completed."
-                )
-
-    def _run_eval(
+    async def _run_inference_eval_pipeline(
         self,
-        inference_job_id: UUID,
-        request: WorkflowCreate,
-        background_tasks: BackgroundTasks,
-        experiment_id: UUID = None,
+        workflow: WorkflowResponse,
+        request: WorkflowCreateRequest,
     ):
+        """Currently this is our only "workflow. As we make this more flexible to handle different
```

Comment:
I think it could pay to start writing this as if we already had some workflow logic. At least, we can separate job creation from job coordination here. WDYT?

Reply:
Totally! I think this ties very well into a comment that @dpoulopoulos made in our meeting today: moving workflow logic into a declarative file so that users can easily build workflows out of the jobs. Created https://github.com/mozilla-ai/lumigator/issues/774 to track this, since I think this totally makes sense to work on after this PR is merged, and can be done in parallel with the other mlflow integration efforts.

Reply:
Ok, let's take it in a separate PR